scripts/proto/lbann_proto.py

```diff
@@ -349,68 +475,73 @@ def __init__(self, **kwargs):
-
-# Generate Callback sub-classes.
+                if type(v) is list: # Repeated field
+                    getattr(callback_message, field_name).extend(v)
+                else: # Singular field
+                    setattr(callback_message, field_name, v)
+        return proto
+
+    # Create sub-class.
+    return type(type_name, (Callback,),
+                {'__init__': __init__, 'export_proto': export_proto})
+
+# Generate Callback sub-classes from lbann.proto
+# Note: The list of skip fields must be updated if any new fields are
+# added to the Callback message in lbann.proto
+_skip_fields = set([])
+_generated_classes = {}
 for field in lbann_pb2.Callback.DESCRIPTOR.fields:
-    _create_proto_callback(field.message_type.name, field.name)
-_add_to_module_namespace(_proto_callbacks)
+    if field.name not in _skip_fields:
+        type_name = field.message_type.name
+        _generated_classes[type_name] = _create_callback_subclass(type_name)
+_add_to_module_namespace(_generated_classes)
 
-def traverse_model(first_module, func):
-    """Traverse a model and apply func to each layer.
+# ==============================================
+# Export models
+# ==============================================
 
-    first_module is the first module in the model (i.e. the input layer).
-    The model is traversed in a depth-first order except that no layer is
-    visited until all its ancestors have been visited.
+def save_model(filename, mini_batch_size, epochs,
+               layers=[], weights=[], objective_function=None,
+               metrics=[], callbacks=[]):
+    """Save a model to file."""
 
-    """
-    visited = set()
-    stack = [first_module]
-    while stack:
-        l = stack.pop()
-        visited.add(l)
-        func(l)
-        for child in l.children:
-            if (child not in visited
-                and all([parent in visited for parent in child.parents])):
-                stack.append(child)
-
-def save_model(filename, first_module, data_layout, mini_batch_size, epochs,
-               objective_func, metrics=[], callbacks=[]):
-    """Save a model to filename.
-
-    Provide the first module (i.e. the input layer) to be saved.
-
-    """
+    # Initialize protobuf message
     pb = lbann_pb2.LbannPB()
-    pb.model.data_layout = data_layout
     pb.model.mini_batch_size = mini_batch_size
     pb.model.block_size = 256  # TODO: Make configurable.
     pb.model.num_epochs = epochs
     pb.model.num_parallel_readers = 0  # TODO: Make configurable
     pb.model.procs_per_model = 0  # TODO: Make configurable
-    # Add the objective function, metrics, and callbacks.
-    objective_func.proto(pb.model)
-    for metric in metrics:
-        metric.proto(pb.model)
-    for callback in callbacks:
-        callback.proto(pb.model)
-    traverse_model(first_module, lambda l: l.proto(pb.model))
+
+    # Add layers
+    layers = list(traverse_layer_graph(layers))
```

Comment:
If we're calling `traverse_layer_graph` here, do we need to pass all the layers in for `layers`? Should document appropriately.